File: Python/src/sort/counting/radix_sort.py
"""
Radix sort for non-negative integers, ported from
sort/counting/RadixSort.java.

One counting sort per decimal digit, least significant first. Each pass is
linear and stable, and stability is what makes the whole thing work: a later
pass may only reorder elements that differ at that digit, leaving the order the
earlier passes established for everything else.

NOTE ``to`` is EXCLUSIVE, as everywhere else in this tree. It used to be
inclusive here, alone among the sorts, so a caller following the usual
convention silently left the last element of the range unsorted. Note also that
only non-negative values work: a negative number gives a negative digit and so a
negative bucket.
"""

from __future__ import annotations

#: Ten buckets, one per decimal digit.
RADIX = 10


def find_max_int(num_arr: list[int], from_: int, to: int) -> int:
    """
    :param num_arr: the values.
    :param from_: the first index to consider.
    :param to: one past the last index to consider, i.e. EXCLUSIVE.
    :return: the largest value in the range.
    """
    return max(num_arr[from_:to])


def count_sort(num_arr: list[int], exp: int, from_: int, to: int) -> None:
    """
    Sort the range by one digit, stably.

    :param num_arr: the values, rearranged in place.
    :param exp: the power of ten selecting the digit: 1, 10, 100, ...
    :param from_: the first index to sort.
    :param to: one past the last index to sort, i.e. EXCLUSIVE.
    """
    result = [0] * len(num_arr)
    count = [0] * RADIX
    for i in range(from_, to):
        count[(num_arr[i] // exp) % RADIX] += 1
    # turn the counts into the position just past each digit's block
    for i in range(1, RADIX):
        count[i] += count[i - 1]
    # NOTE backwards, which is what makes the pass stable: the last element with
    # a given digit is placed last, so equal digits keep the order they were in.
    for i in range(to - 1, from_ - 1, -1):
        digit = (num_arr[i] // exp) % RADIX
        count[digit] -= 1
        result[count[digit] + from_] = num_arr[i]
    if to - from_ >= 0:
        num_arr[from_:to] = result[from_:to]


def sort(num_arr: list[int], from_: int, to: int) -> None:
    """
    Sort the range, one decimal digit at a time.

    :param num_arr: the values, rearranged in place.
    :param from_: the first index to sort.
    :param to: one past the last index to sort, i.e. EXCLUSIVE.
    :raises ValueError: if from_ is greater than to.
    :raises IndexError: if either index is outside the list.
    """
    if num_arr is None or len(num_arr) == 1 or from_ == to:
        return
    if from_ > to:
        raise ValueError("From value should be less than to")
    if from_ < 0 or from_ > len(num_arr) - 1:
        raise IndexError(f"From should be between 0 and {len(num_arr) - 1}")
    if to > len(num_arr):
        raise IndexError(f"To should be between 0 and {len(num_arr)}")
    max_val = find_max_int(num_arr, from_, to)
    exp = 1
    while max_val // exp > 0:
        count_sort(num_arr, exp, from_, to)
        exp *= RADIX

File: Java/src/main/java/com/phasmidsoftware/dsaipg/sort/counting/RadixSort.java
package com.phasmidsoftware.dsaipg.sort.counting;

/**
 * TODO eliminate this class (and its tests)
 *
 * Radix Sort
 * Radix sort is an integer sorting algorithm that sorts data with integer keys
 * by grouping the keys by individual digits that share the same significant
 * position and value (place value). Radix sort uses counting sort as a
 * subroutine to sort an array of numbers.
 * <p>TESTME
 * <b>Disclaimer:</b> This radix sort can only sort positive integers
 *
 * @version 1.0
 * @since 13th May 2020
 */
public class RadixSort {

    /**
     * findMaxInt method is used to find maximum number in the array
     * within the provided range i.e from and to
     *
     * @param numArr It contains an array of numbers from which maximum value needs to be obtained
     * @param from   This is the starting index from which scanning for maximum number will begin
     * @param to     This is the ending index until which scanning for maximum number will be continued
     * @return int This method returns maximum number between from index and to index
     */
    /**
     * @param numArr the values.
     * @param from   the index of the first value to consider.
     * @param to     one past the index of the last, i.e. EXCLUSIVE.
     * @return the largest value in the range.
     */
    public int findMaxInt(int[] numArr, int from, int to) {
        int maxVal = numArr[from];
        for (int i = from; i < to; i++)
            maxVal = Math.max(maxVal, numArr[i]);
        return maxVal;
    }

    /**
     * countSort method is implementation of basic counting sort algorithm.
     * We provide exponent i.e unit's digit, ten's digit or hundred's digit etc on which counting sort needs to be performed
     *
     * @param numArr It contains an array of numbers on which counting sort needs to be performed
     * @param exp    This is the exponent input on which counting sort would be performed e.g 1, 10, 100, 1000 etc.
     * @param from   This is the starting index from which sorting operation will begin
     * @param to     This is the ending index until which sorting operation will be continued
     */
    /**
     * Sort the range by one digit, stably. Stability is the mechanism: a later
     * pass may only reorder values that differ at that digit, so the order the
     * earlier passes established survives.
     *
     * @param numArr the values, rearranged in place.
     * @param exp    the power of ten selecting the digit: 1, 10, 100, ...
     * @param from   the index of the first value to sort.
     * @param to     one past the index of the last, i.e. EXCLUSIVE.
     */
    public void countSort(int[] numArr, int exp, int from, int to) {
        int arrLength = numArr.length;
        int[] result = new int[arrLength]; //This stores output result
        int[] count = new int[10]; // This maintains digit wise occurrence count

        //This method records occurrence of digits in count[]
        for (int i = from; i < to; i++)
            count[(numArr[i] / exp) % 10]++;

        // Modifying value of count[i] so that it now contains actual position of the digit
        for (int i = 1; i < 10; i++)
            count[i] += count[i - 1];

        // Building result array to contain radix sorted output array on selected exponent
        for (int i = to - 1; i >= from; i--) {
            result[count[(numArr[i] / exp) % 10] - 1 + from] = numArr[i];
            count[(numArr[i] / exp) % 10]--;  // Reducing count[] to adjust the next location of particular digit
        }

        //Copying result array in original array
        if (to - from >= 0) System.arraycopy(result, from, numArr, from, to - from);
    }

    /**
     * sort method is implementation of radix sort algorithm.
     *
     * @param numArr It contains an array of numbers on which radix sort needs to be performed
     * @param from   This is the starting index from which sorting operation will begin
     * @param to     This is the ending index until which sorting operation will be continued
     */
    /**
     * Sort the range, one decimal digit at a time, least significant first.
     * <p>
     * NOTE {@code to} is EXCLUSIVE, as everywhere else in this tree. It used to be
     * inclusive here, alone among the sorts, so a caller following the usual
     * convention silently left the last element of the range unsorted.
     * <p>
     * NOTE only non-negative values work: a negative value yields a negative digit
     * and so a negative bucket index.
     *
     * @param numArr the values, rearranged in place.
     * @param from   the index of the first value to sort.
     * @param to     one past the index of the last, i.e. EXCLUSIVE.
     * @throws Exception if from is greater than to.
     */
    public void sort(int[] numArr, int from, int to) throws Exception {

        // Sort Validations on input

        if (numArr == null || numArr.length == 1 || from == to) return;

        if (from > to) throw new Exception("From value should be less than to");

        if (from < 0 || (from > numArr.length - 1))
            throw new ArrayIndexOutOfBoundsException("From should be between 0 and " + (numArr.length - 1));

        if (to > numArr.length)
            throw new ArrayIndexOutOfBoundsException("To should be between 0 and " + numArr.length);

        // Finding max number
        int maxVal = findMaxInt(numArr, from, to);
        //Performing counting sort on every exponent
        int exp = 1;
        while (maxVal / exp > 0) {
            countSort(numArr, exp, from, to);
            exp *= 10;
        }
    }

}
